### ISSUE TITLE
[22.0.0] InternalServer when we try to impersonate

### ISSUE THREAD

**Issue by vbdalmaz** [keycloak/keycloak#21654]
After updating to 22.0 we are receiving an internal server error when we try to impersonate:

We found the error Caused by: freemarker.template.TemplateNotFoundException: Template not found for name "index.ftl".

We customized the template /realms/realmsIDName/account/ but never had index.ftl inside the account folder on templates.

Maybe it's an issue, moving to issue

**Comment by jonkoops** [comment]
@vbdalmaz could this possibly be because you have 'base' theme configured? I've seen a similar issue with the Admin Console (see #21421).

**Comment by vbdalmaz** [comment]
![Screen Shot 2023-07-12 at 13 04 02](https://github.com/keycloak/keycloak/assets/2130513/d521297f-b3d4-4901-b0fe-b08ddb38ae91)

Before 21.1.1 was working fine. I am building a new one to try

**Comment by vbdalmaz** [comment]
![Screen Shot 2023-07-12 at 13 05 45](https://github.com/keycloak/keycloak/assets/2130513/c4c1d33b-9542-40f5-9fb3-ac991f8079ad)

I have @jonkoops thanks for answering. On previous 21.1.1 was working fine.

**Comment by phroust** [comment]
I have encountered the same problem.

We are using custom theming for most pages (Acount, Email, Login and Welcome). All themed pages except Account page work in Keycloak 22.

The Account theme breaks in Keycloak 22 due to missing "index.ftl". Our account theme never hat an "index.ftl"

We have configured `parent=base` in `theme.properties`.

**Comment by jonkoops** [comment]
Yeah, this is definitely a high impact issue that needs to be resolved. I'm marking this for the next patch release.

**Comment by pedroigor** [comment]
Just tested this issue and it seems happening since 21. If you run a `20.0.5` instance and change the Admin Console theme to `base` you won't get an exception but a non-styled page (as it was to be in legacy too).

**Comment by vbdalmaz** [comment]
Just another update, He throws an Internal Server error but if you access another page it works, so because of this I am lowering the impact.

**Comment by phroust** [comment]
@pedroigor: yes, using the base theme itself is broken since Keycloak 21. However, using the base theme as the parent for a custom theme worked fine until Keycloak 22.

Using `parent=base` for a custom theme is the official way documented [here](https://www.keycloak.org/docs/latest/server_development/#creating-a-theme), but this doesn't work anymore for Admin and Account type themes.

**Comment by jonkoops** [comment]
Adding this to the next patch release, so that it will not be forgotten.

**Comment by mohamedbasuney** [comment]
Inside keycloak 22.0.1 themes directory I can add these directories (email, login, welcome, common, account) when add admin folder I get this error Template not found for name "index.ftl".

**Comment by micahalgard** [comment]
Has anyone found a workaround for this? Still seems to be an issue in the nightly build.

**Comment by stianst** [comment]
The reason this stopped working in 22 is the removal of Account Console v1, which is at the same time stopping themes using the older account console from working.

A workaround that should work for now is to update the custom account theme to work with the newer Account Console v2, and use `parent=keycloak.v2` instead of `parent=base`.

I think it would make sense to introduce a `base` theme for Account Console v2 as well, that is basically functional, but without any styling, while `keycloak.v2` adds the standard PatternFly styling only.

**Comment by jonkoops** [comment]
I am starting to think we should get rid of the base themes in general. A lot of logic can be inside of them that is not necessarily useful for the themes extending them. Ideally I would like to stop extending the `base` theme for the V2 of the Admin Console and V3 of the Account Console.

**Comment by jonkoops** [comment]
Moving this to v23, as we will no longer be doing patch releases for v22.

**Comment by jonkoops** [comment]
Since this has not been picked up and we're doing a release, I am moving this to the next version.

**Comment by hmlnarik** [comment]
Summarizing the comments above and related issues, e.g. #22671:
- `base` theme is [used to contain unstyled account console v1](https://github.com/keycloak/keycloak/tree/release/21.1/themes/src/main/resources/theme/base) and does not anymore in main since #21098
- `base` theme still contains message keys which are used in account console 2, e.g. `accountManagementTitle`. Account console 3 has all keys covered in its own message bundle.
  <details>
  <summary>Full list of account v2 missing keys</summary>
  <ul>
  <li>account</li>
  <li>accountManagementTitle</li>
  <li>accountSecurityIntroMessage</li>
  <li>accountSecuritySidebarTitle</li>
  <li>accountUpdatedMessage</li>
  <li>application</li>
  <li>applicationName</li>
  <li>applications</li>
  <li>applicationsIntroMessage</li>
  <li>backTo</li>
  <li>deviceActivitySidebarTitle</li>
  <li>doCancel</li>
  <li>doLogIn</li>
  <li>doSave</li>
  <li>doSignOut</li>
  <li>email</li>
  <li>filterByName</li>
  <li>firstName</li>
  <li>icon</li>
  <li>ip</li>
  <li>lastName</li>
  <li>linkedAccountsSidebarTitle</li>
  <li>myResources</li>
  <li>name</li>
  <li>noResourcesSharedWithYou</li>
  <li>notHaveAnyResource</li>
  <li>personalInfoIntroMessage</li>
  <li>personalInfoSidebarTitle</li>
  <li>personalSubMessage</li>
  <li>requestor</li>
  <li>resourceIntroMessage</li>
  <li>resources</li>
  <li>sessions</li>
  <li>sharedwithMe</li>
  <li>signingInSidebarTitle</li>
  <li>username</li>
  </ul>
  </details>
  <details>
  <summary>How to obtain the potential missing keys</summary>
        <tt>cd themes/src/main/resources/theme/</tt><br>
        <tt>KEYS_BASE_MINUS_V2=$(comm -23 <(perl -ne 'print "$1\n" if (/^(\w+)=/)' base/account/messages/messages_en.properties | sort) <(perl -ne 'print "$1\n" if (/^(\w+)=/)' keycloak.v2/account/messages/messages_en.properties | sort))</tt><br>
        <tt>for I in $KEYS_BASE_MINUS_V2; do if git grep -qE "[\"']$I[\"']" keycloak.v2/account/; then echo "$I"; fi; done</tt>
  </details>
- `base` theme is referred to in the [documentation](https://github.com/keycloak/keycloak/blob/release/23.0/docs/documentation/server_development/topics/themes.adoc) which shows the way of extending themes

I can see two options:
1. Remove the base theme completely:
   - `account.v2` will become a top-level theme
   - Any message key that is inherited from `base` will be transferred to `account.v2`
   - Documentation will be updated to refer to a parent theme rather than specifying `base` as an example
   - The cost is that with development of another account theme (e.g. account.v3), one would need to maintain two sets of potentially overlapping translations (we're in that situation already with [v2](https://github.com/keycloak/keycloak/tree/main/themes/src/main/resources/theme/keycloak.v2/account/messages) and [v3](https://github.com/keycloak/keycloak/tree/main/js/apps/account-ui/maven-resources/theme/keycloak.v3/account/messages))
2. Keep the base theme as the base for the translations and mandate creation of `index.ftl` in custom themes
   - When `index.ftl` would not be found in the theme, a simple text page would be shown instead stating `You need to supply index.ftl in your theme, please see documentation`
   - Documentation will be updated to describe the requirements - existence of `index.ftl`
   - This would allow sharing translations between the themes

It seems Keycloak codebase is inclining towards 1. Happy to hear arguments or preference to following either option.

**Comment by jonkoops** [comment]
@hmlnarik My vote is for option 1, remove all the `base` themes. Our themes are already strangely fractured, yet somehow dependent on each other, let's trim the fat. Also, I eventually want to get rid of all versioned themes, and simply have a single `keycloak` theme containing all of them, this only exists for historical reasons, and adds extra context that is not needed for new and most existing users.

**Comment by ssilvert** [comment]
Many may not know this, but our REST APIs use message bundles stored in the themes.  This is obviously wrong as the REST API should have no dependency on themes or even the realm.  These messages should stored somewhere that they can be accessed globally.  I can't think of a use case where you would want a server-side error message to care about what theme or realm is active.  If you needed, for instance, the realm name, you would just pass it into an interpolated message.

This was the motivation for https://github.com/keycloak/keycloak/issues/22671

**Comment by ssilvert** [comment]
Other thoughts:

- I agree that we need to eventually get rid of the versioned themes if we can.
- It will be awhile before the base theme goes away because it is so essential for custom versions of login.
- No matter what, I don't want to see multiple copies of the same message bundles.

**Comment by jonkoops** [comment]
@hmlnarik I think we can close this one? Seems like this was resolved under #26596

**Comment by parvuselephantus** [comment]
Was it? I'm running 24.0.1 and getting error: " Uncaught server error: org.keycloak.theme.FreeMarkerException: Failed to process template index.ftl" while trying to access realm/<realm-id>/account

**Comment by hmlnarik** [comment]
@parvuselephantus Can you share details of the template you use for account console:
- Which theme do you use? 
- Is it a built-in or customized in any way?
- If customized, what is its parent theme?

**Comment by parvuselephantus** [comment]
I ran out of time and don't have exactly what I used to have. However I started from 20.0.5 environment to upgrade into 24.0.1. I used a customized theme, which was a modified base theme. The parent=base. Later on I played with versions of parent=keycloak* - some worked some didn't. I'm not any great on keycloak - not sure when template is applied and when not (when the restart is needed) - played also with index.ftl being copy of accoun.ftl or just empty file - interesting in some cases (can't remember now which) it started fine, but even though eg having index.ftl empty I saw the account page with the content that seems to me to be the default (one of keycloak* versions).
From the discussion I understand now the 'parent=' idea is planned to be scrapped and I should rather copy full theme and change it (instead of copying only modified files). Just wanted to mention, that if you go this path, remember that I will need to take care of multiple files I never modified (if you change variable naming convention). Keeping fingers crossed for you all, to accomplish full backward compatibility :)

**Comment by Lisser** [comment]
I ran into this yesterday when creating a custom theme. I resolved the problem, but after reading through this issue I am left wondering what the current suggested way of implementing and maintaining a custom theme is. 

I currently implemented my theme as following the instructions in the documentation repo. E.g.:

```
parent=base
import=common/keycloak
styles=...
```

However, is it better to copy the whole `base`-theme and make changes, instead of inheriting from it?

**Comment by parvuselephantus** [comment]
@Lisser - that will not work for you long term, because Keycloak seems to change base often. So variable markers will get obsolete on next update and you will have to update lots of code you have never written (just copied). If I recall (sorry - it was over a month and with such a volume of issues I don't remember each one) the way to go for me was to use parent=kecyloak3 (I think keycloak 3 was the newest month ago). Copy only files you want to alter - the other should be updated as Keycloak updates are released (you will need to reimplement files you changed on some updates later on). Also if I recall there was idea to make parent= setting obsolete... but we never know which path will Keycloak go.

**Comment by Lisser** [comment]
Alright! I'll keep my extension of base for now, it's not much and should be easily ported to a new realease. 

I'm curious to hear the official stance on how theme development should be done in the future. If you need help rewriting the documentation at any point, don't hesitate to reach out @hmlnarik
